### PR TITLE
Extract a Pane base separate from a MonacoPane.

### PR DESCRIPTION
**Pull request by mattgodbolt**
Paves the way to move the (few) tools that don't have an editor
to also share the base class.

### CODE REVIEW

**Comment by mattgodbolt** [comment]
Dipping my toes in typescript here - please be gentle but thorough!! :)

The intention is to be able to phrase things like the CFG in terms of this.

File: static/panes/pane.interfaces.ts
Comment (partouf) [inline static/panes/pane.interfaces.ts:33]: I think especially when overriding, and its still just data, this should probably be a class and not an interface

Comment (partouf) [reply static/panes/pane.interfaces.ts:33]: Same with MonacoPaneState

File: static/panes/pane.interfaces.ts
Comment (partouf) [inline static/panes/pane.interfaces.ts:36]: shouldn't this be `editorId` (uppercase i) ?

File: static/panes/pane.ts
Comment (partouf) [inline static/panes/pane.ts:41]: ah it's defined here in a different way.. this is really confusing

**Review by partouf — APPROVED** [review]

File: static/panes/pane.ts
Comment (matsjla) [inline static/panes/pane.ts:268]: I don't really like this type assertion because it's always going to be wrong. I think a better solution is
```suggestion
    getCurrentState(): MonacoPaneState {
        const parent = super.getCurrentState();
        const state: MonacoPaneState = {
            selection: this.selection,
            ...parent,
        };
        this.fontScale.addState(state);
        return state;
    }
```
(assuming this would compile, just did github suggestion)

**Review by matsjla — CHANGES_REQUESTED** [review]
Looks very good overall!

File: static/panes/pane.interfaces.ts
Comment (mattgodbolt) [reply static/panes/pane.interfaces.ts:33]: Works for me. It was an `interface` before so I kept that but didn't understand the difference.

File: static/panes/pane.ts
Comment (mattgodbolt) [reply static/panes/pane.ts:268]: That's what I had to start with I think -- I'll check though!

File: static/panes/pane.interfaces.ts
Comment (mattgodbolt) [reply static/panes/pane.interfaces.ts:33]: is it OK to have `interfaces` in the `interfaces.ts` file?

File: static/panes/pane.ts
Comment (mattgodbolt) [reply static/panes/pane.ts:272]: That works! thanks!

**Comment by mattgodbolt** [comment]
Will hold off merging til after @jeremy-rifkin 's PR gets reviewed and merged. There's going to be some minor conflicts and I'd prefer not to inflict that on someone else :)

**Comment by mattgodbolt** [comment]
So I just also made all the `number | boolean` fields optional numbers instead: I think this is better as:
* Old states in URLs will have "missing" `treeid`s anyway, so `number?` is a better model of that
* It seems _slightly_ less confusing than having to ask "what does `editorid === true` mean?

Thoughts welcomed though, especiailly from @jeremy-rifkin 

I got as far as a quick test but before this can be merged I'll need to do some testing (including with some existing URLs)

File: static/panes/ast-view.js
Comment (jeremy-rifkin) [inline static/panes/ast-view.js:153]: I believe this will cause an issue with the "Editor #false" bug, the UI code does actually pass `false` rather than `undefined` when in tree mode

File: static/panes/cfg-view.interfaces.ts
Comment (jeremy-rifkin) [inline static/panes/cfg-view.interfaces.ts:28]: Good point about existing URLs that don't have `treeid`. If I remember correctly, the treeid can also be a `false` value from the other UI code so I think `treeid?: number | boolean` would probably be most correct.

`treeid?: number` (and same with `editorid`) would probably be better overall, not sure there's an easy way to address the source of the `false` id but they could be eliminated here https://github.com/compiler-explorer/compiler-explorer/blob/17b4f2ebc426c7d183a67a4a49b82c28640bef39/static/components.js#L158-L447

Comment (mattgodbolt) [reply static/panes/cfg-view.interfaces.ts:28]: > from the other UI code

Is this UI code already released on the site? Or just stuff that was added as part of the earlier PR? If the latter we have time to update and change that behaviour.

Comment (mattgodbolt) [reply static/panes/cfg-view.interfaces.ts:28]: > from the other UI code

Is this UI code already released on the site? Or just stuff that was added as part of the earlier PR? If the latter we have time to update and change that behaviour.

File: static/panes/ast-view.js
Comment (mattgodbolt) [reply static/panes/ast-view.js:153]: As discussed in Discord I'm updating these to all be `if (this._editorid)` etc - please take another look when you get a chance! Thanks again for your patience whlie I learn all this new stuff :-)

File: static/panes/cfg-view.interfaces.ts
Comment (mattgodbolt) [reply static/panes/cfg-view.interfaces.ts:28]: I _think_ my recent checkin deals with this. Please take a look again! Thanks!

Comment (jeremy-rifkin) [reply static/panes/cfg-view.interfaces.ts:28]: From earlier code / current site behavior yeah, not sure the origin exactly!

**Comment by mattgodbolt** [comment]
Apologies for all the review request, I'm utterly paranoid I'm doing something wrong and trying to build up my confidence in TS :)

**Comment by mattgodbolt** [comment]
This is now live

**Pull request merged**
